plugins/woocommerce/src/Blocks/Domain/Services/CheckoutLink.php

```diff
@@ -0,0 +1,128 @@
+<?php
+/**
+ * Functionality that takes a static URL, constructs a cart, and redirects to the checkout with a cart session.
+ */
+
+declare(strict_types=1);
+
+namespace Automattic\WooCommerce\Blocks\Domain\Services;
+
+use Automattic\WooCommerce\Blocks\Utils\CartCheckoutUtils;
+use Automattic\WooCommerce\StoreApi\Utilities\CartTokenUtils;
+use Automattic\WooCommerce\StoreApi\Utilities\CartController;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Checkout Link class.
+ */
+class CheckoutLink {
+	/**
+	 * Initialize the checkout link service.
+	 */
+	public function init() {
+		add_action( 'init', array( $this, 'add_checkout_link_endpoint' ) );
+		add_filter( 'query_vars', array( $this, 'add_query_vars' ), 0 );
+		add_action( 'template_redirect', array( $this, 'handle_checkout_link_endpoint' ) );
+	}
+
+	/**
+	 * Add the checkout link endpoint.
+	 */
+	public function add_checkout_link_endpoint() {
+		add_rewrite_rule( '^checkout-link$', 'index.php?checkout-link=true', 'top' );
+	}
+
+	/**
+	 * Add the checkout link query var.
+	 *
+	 * @param array $vars The query vars.
+	 * @return array The query vars.
+	 */
+	public function add_query_vars( $vars ) {
+		$vars[] = 'checkout-link';
+		return $vars;
+	}
+
+	/**
+	 * Handle the checkout link endpoint.
+	 *
+	 * @return void
+	 */
+	public function handle_checkout_link_endpoint() {
+		if ( ! get_query_var( 'checkout-link' ) ) {
+			return;
+		}
+		wp_safe_redirect( $this->get_checkout_link() );
+		exit;
+	}
+
+	/**
+	 * Process the query params and return the checkout link to redirect to complete with session token.
+	 *
+	 * @return string The checkout link.
+	 */
+	protected function get_checkout_link() {
+		$controller = new CartController();
+		$controller->empty_cart();
+
+		// Populate cart with products.
+		$products = array_filter( explode( ',', wc_clean( wp_unslash( $_GET['products'] ?? '' ) ) ) ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+
+		foreach ( $products as $product_id_qty ) {
+			if ( strpos( $product_id_qty, ':' ) !== false ) {
+				list( $product_id, $qty ) = explode( ':', $product_id_qty );
+			} else {
+				$product_id = $product_id_qty;
+				$qty        = 1;
+			}
+			if ( ! absint( $product_id ) ) {
+				continue;
+			}
+			try {
+				$controller->add_to_cart(
+					[
+						'id'       => absint( $product_id ),
+						'quantity' => absint( $qty ),
+					]
+				);
+			} catch ( \Exception $e ) {
+				wc_add_notice( $e->getMessage(), 'error' );
+			}
+		}
+
+		// Apply coupon if provided.
+		$coupon = wc_format_coupon_code( wp_unslash( $_GET['coupon'] ?? '' ) ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended, WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+
+		if ( wc_coupons_enabled() && ! empty( $coupon ) ) {
+			try {
+				$controller->apply_coupon( $coupon );
```

Comment:
Can you add coupons to an empty cart? in my testing that wasn't possible, not providing products or providing invalid products ended up not adding the coupon, but it did't effectively empty the cart, adding a product manually would surface the coupon.

Reply:
We can swap it so it adds coupon after checking cart is empty. Done in https://github.com/woocommerce/woocommerce/pull/58140/commits/18d7493cd40885ea849cfeb261c86a100bd8614b